core/accounting.go

```diff
@@ -50,6 +95,20 @@ func (b *Balances) Debit(id ManifestID, amount *big.Rat) {
 	b.balances[id].lastUpdate = time.Now()
 }
 
+// Reserve zeros the balance for a ManifestID and returns the current balance
+func (b *Balances) Reserve(id ManifestID) *big.Rat {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	if b.balances[id] == nil {
+		b.balances[id] = &balance{amount: big.NewRat(0, 1)}
+	}
+
+	amount := b.balances[id].amount
+	b.balances[id].amount = big.NewRat(0, 1)
```

Comment:
We're setting the balance to zero here with the expectation that it might become non-zero later on (i.e. crediting positive change back to the balance) so I don't think we want to delete the key. Deleting the key would indicate that the entry is not present in the map at all (since `delete` would set `balance = nil`) as opposed to indicating that the entry is present with a balance of zero. Another example of setting the balance to zero without wanting to delete the key is when a caller invokes `Debit()` such that after subtracting the provided amount the balance becomes zero.

Reply:
Yeah I entirely agree with that reasoning. 

It's just that both `Credit` and `Debit` make no distinction between `balance` nil and `balance.amount = 0` since they zero the balance for a non-existing key before executing any logic.

So I guess depending on the frequency of expected non-zero updates later on there is a tradeoff here between less `Cleanup` work and having to do a re-allocation when updating deleted key.  Not that it matters that much and it logically checks out more to have a zeroed balance in this case rather than a deleted key.